File: dbflow-sqlcipher/src/main/java/com/raizlabs/android/dbflow/sqlcipher/SQLCipherStatement.java
package com.raizlabs.android.dbflow.sqlcipher;

import com.raizlabs.android.dbflow.structure.database.BaseDatabaseStatement;
import com.raizlabs.android.dbflow.structure.database.DatabaseStatement;

import net.sqlcipher.database.SQLiteStatement;

/**
 * Description: Implements the methods necessary for {@link DatabaseStatement}. Delegates calls to
 * the contained {@link SQLiteStatement}.
 */
public class SQLCipherStatement extends BaseDatabaseStatement {

    public static SQLCipherStatement from(SQLiteStatement statement) {
        return new SQLCipherStatement(statement);
    }

    private final SQLiteStatement statement;

    SQLCipherStatement(SQLiteStatement statement) {
        this.statement = statement;
    }

    public SQLiteStatement getStatement() {
        return statement;
    }

    @Override
    public long executeUpdateDelete() {
        return statement.executeUpdateDelete();
    }

    @Override
    public void execute() {
        statement.execute();
    }

    @Override
    public void close() {
        statement.close();
    }

    @Override
    public long simpleQueryForLong() {
        return statement.simpleQueryForLong();
    }

    @Override
    public String simpleQueryForString() {
        return statement.simpleQueryForString();
    }

    @Override
    public long executeInsert() {
        return statement.executeInsert();
    }

    @Override
    public void bindString(int index, String s) {
        statement.bindString(index, s);
    }

    @Override
    public void bindNull(int index) {
        statement.bindNull(index);
    }

    @Override
    public void bindLong(int index, long aLong) {
        statement.bindLong(index, aLong);
    }

    @Override
    public void bindDouble(int index, double aDouble) {
        statement.bindDouble(index, aDouble);
    }

    @Override
    public void bindBlob(int index, byte[] bytes) {
        statement.bindBlob(index, bytes);
    }
}

File: sqlcipher/src/main/kotlin/com/dbflow5/sqlcipher/SQLCipherStatement.kt
package com.dbflow5.sqlcipher

import com.dbflow5.database.DatabaseStatement
import net.sqlcipher.database.SQLiteStatement

/**
 * Description: Implements the methods necessary for [DatabaseStatement]. Delegates calls to
 * the contained [SQLiteStatement].
 */
class SQLCipherStatement
internal constructor(val statement: SQLiteStatement) : DatabaseStatement {

    override fun executeUpdateDelete(): Long =
        rethrowDBFlowException { statement.executeUpdateDelete().toLong() }

    override fun execute() {
        statement.execute()
    }

    override fun close() {
        statement.close()
    }

    override fun simpleQueryForLong(): Long =
        rethrowDBFlowException { statement.simpleQueryForLong() }

    override fun simpleQueryForString(): String? =
        rethrowDBFlowException { statement.simpleQueryForString() }

    override fun executeInsert(): Long = rethrowDBFlowException { statement.executeInsert() }

    override fun bindString(index: Int, s: String) {
        statement.bindString(index, s)
    }

    override fun bindNull(index: Int) {
        statement.bindNull(index)
    }

    override fun bindLong(index: Int, aLong: Long) {
        statement.bindLong(index, aLong)
    }

    override fun bindDouble(index: Int, aDouble: Double) {
        statement.bindDouble(index, aDouble)
    }

    override fun bindBlob(index: Int, bytes: ByteArray) {
        statement.bindBlob(index, bytes)
    }

    override fun bindAllArgsAsStrings(selectionArgs: Array<String>?) {
        if (selectionArgs != null) {
            for (i in selectionArgs.size downTo 1) {
                bindString(i, selectionArgs[i - 1])
            }
        }
    }

    companion object {

        @JvmStatic
        fun from(statement: SQLiteStatement): SQLCipherStatement = SQLCipherStatement(statement)
    }
}
